addon/adapters/osf-adapter.js

```diff
@@ -18,97 +24,199 @@ export default DS.JSONAPIAdapter.extend(HasManyQuery.RESTAdapterMixin, DataAdapt
+    },
+    _deleteRelated(store, snapshot, removedSnapshots) { // jshint ignore:line
+        return this._removeRelated(...arguments).then(() => {
+            if (removedSnapshots.record) {
+                removedSnapshots = [removedSnapshots];
+            }
+            removedSnapshots.forEach(s => s.record.unloadRecord());
+        });
+    },
+    _doRelatedRequest(store, snapshot, relatedSnapshots, relationship, url, requestMethod, isBulk = false) {
+        var data = {};
+        var relatedMeta = snapshot.record[relationship].meta();
+        var type = singularize(relatedMeta.type);
+        var serializer = store.serializerFor(type);
+        if (relatedSnapshots.record) {
+            serializer.serializeIntoHash(
+                data,
+                store.modelFor(type),
+                relatedSnapshots, {
+                    forRelationship: true,
+                    isBulk: isBulk
+                }
+            );
         } else {
-            serialized = serialized[0];
+            data.data = relatedSnapshots.map(relatedSnapshot => {
+                var item = {};
+                serializer.serializeIntoHash(
+                    item,
+                    store.modelFor(type),
+                    relatedSnapshot, {
+                        forRelationship: true,
+                        isBulk: isBulk
+                    }
+                );
+                return item.data;
+            });
         }
-        return serialized;
+        return this.ajax(url, requestMethod, {
+            data: data,
+            isBulk: isBulk
+        }).then(res => {
+            if (!res) {
+                if (requestMethod === 'DELETE') {
+                    res = [];
+                } else {
+                    return null;
+                }
+            }
+            return res;
+        });
     },
-    _handleManyRequest(store, type, snapshot, query, relationship, serializer) {
-        var relationMeta = snapshot.record[relationship].meta();
-        var serialized = snapshot.hasMany(relationship)
-            .filter(each => each.record.isNewOrDirty())
-            .map(each => serializer.serialize(each));
-        var options = Ember.merge({
-                requestType: () => 'PATCH',
-                isBulk: serialized => serialized.length > 1,
-                url: this._buildRelationshipURL(type.modelName, snapshot.id, snapshot, 'updateRecord', query, relationship),
-                serialized: this._serializeHasMany
-            },
-            relationMeta.options.updateRequest
-        );
-        return this.ajax(options.url, options.requestType(snapshot, relationship), {
-            data: options.serialized(serialized),
-            isBulk: options.isBulk(serialized)
-        }).then(() => snapshot.record.clearDirtyRelationship(relationMeta));
+    _handleRelatedRequest(store, type, snapshot, relationship, change) {
```

Comment:
Some documentation (especially of the change parameter and allowed values) would be extremely helpful for maintainability here. :)